base/some.jl

```diff
@@ -145,3 +145,5 @@ macro something(args...)
     something = GlobalRef(Base, :something)
     return :($something($expr))
 end
+
```

Comment:
```suggestion

hash(s::Some) = hash(s.value)
```
`hash` needs to obey `x == y` => `hash(x) == hash(y)`, so I believe this change also needs to modify `hash`.

Reply:
Hmm, I'm not sure if it's right that `hash(Some(1)) == hash(1)` then. Maybe `hash(s::Some) = hash(s.value, hash(Some))` would be a better choice?

Reply:
Good catch!

Reply:
Good point @tecosaur. Maybe to make sure we don't waste time doing `hash(Some)` (takes around 6ns on my machine because `Some` is mutable, it can't constant fold) we can compute it ahead of time? e.g. something like
```suggestion

const _hashSome = hash(gensym(:Some))
hash(s::Some) = hash(_hashSome, s.value)
```

Reply:
It won't constant fold either way because you need to implement `hash(s::Some, h::UInt)` and incorporate that `h`.

Reply:
Pity about that, seems like reasonable bandaid though.

Reply:
gotcha.

Reply:
What about `hash(s::Some, h::UInt) = hash(s.value, hash(_hashSome, h))`? Seems faster on my machine.

Reply:
Yeah seems profitable on my machine too:
```julia
julia> @eval Base hash(s::Some, h::UInt) = hash(s.value, hash(Some, h))
hash (generic function with 75 methods)

julia> let x = Ref(Some(1))
           @btime hash($x[])
       end
  6.622 ns (0 allocations: 0 bytes)
0xdccec56b562d01aa

julia> @eval Base begin
       const _hashSome = hash(Some)
       hash(s::Some, h::UInt) = hash(s.value, hash(_hashSome, h))
       end
hash (generic function with 75 methods)

julia> let x = Ref(Some(1))
           @btime hash($x[])
       end
  2.164 ns (0 allocations: 0 bytes)
0xc132aa77a9d9ec06
```

Reply:
I don't think this will make a difference in practice:

```
julia> @benchmark hash(s) setup=(s=Some(rand(UInt))) evals=1
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.000 ns … 410.000 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     20.000 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   19.896 ns ±   4.446 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▃                            █                               ▁
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ █
  10 ns         Histogram: log(frequency) by time        30 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @eval Base begin
       const _hashSome = hash(Some)
       hash(s::Some, h::UInt) = hash(s.value, hash(_hashSome, h))
       end
hash (generic function with 61 methods)

julia> @benchmark hash(s) setup=(s=Some(rand(UInt))) evals=1
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.000 ns … 300.000 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     20.000 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   18.026 ns ±   5.048 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                                                             █  
  ▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁▁█ ▂
  10 ns           Histogram: frequency by time           20 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

It's very unlikely that users will hash the exact same `Some` object over and over and over again.

Reply:
That's just because you're using `evals=1` which is inaccurate for such a fast function. 
```julia
julia> @eval Base hash(s::Some, h::UInt) = hash(s.value, hash(Some, h))
hash (generic function with 75 methods)

julia> @benchmark hash(s) setup=(s=Some(rand(UInt)))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  6.041 ns … 24.126 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     6.693 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.700 ns ±  0.262 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▁                                                       ▅█ ▁
  ██▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁██ █
  6.04 ns      Histogram: log(frequency) by time      6.7 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @eval Base begin
       const _hashSome = hash(Some)
       hash(s::Some, h::UInt) = hash(s.value, hash(_hashSome, h))
       end
hash (generic function with 75 methods)

julia> @benchmark hash(s) setup=(s=Some(rand(UInt)))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.374 ns … 6.934 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     3.036 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   3.226 ns ± 0.456 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                              █                           ▇  
  ▆▂▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁▁▁▂▁▁▁▁▁▁▁▁█▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▂
  2.37 ns        Histogram: frequency by time       3.68 ns <

```
vs
```julia
julia> @eval Base begin
       const _hashSome = hash(Some)
       hash(s::Some, h::UInt) = hash(s.value, hash(_hashSome, h))
       end
hash (generic function with 75 methods)

julia> @benchmark hash(s) setup=(s=Some(rand(UInt)))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.374 ns … 6.934 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     3.036 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   3.226 ns ± 0.456 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                              █                           ▇  
  ▆▂▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁▁▁▂▁▁▁▁▁▁▁▁█▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▂
  2.37 ns        Histogram: frequency by time       3.68 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

```

Reply:
You can make any two functions that take less than 10 or so nanoseconds look like they take the same amount of time by using `evals=1`.

Reply:
e.g.
```julia
julia> @benchmark prod(v) setup=(v=rand(5)) evals = 1
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  20.000 ns … 281.000 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     20.000 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   21.978 ns ±   4.848 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                             
  █▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▁▁▁▂ ▂
  20 ns           Histogram: frequency by time           31 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark prod(v) setup=(v=rand(10)) evals = 1
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  20.000 ns … 291.000 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     20.000 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   21.842 ns ±   4.807 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █    ▃                                                ▆      ▁
  █▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▇ █
  20 ns         Histogram: log(frequency) by time        31 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Reply:
That says more about the flakiness of these nanobenchmarks than about the added utility of the constant.

I'll maintain that the current version in this PR is perfectly fine, and should this become a performance bottleneck in the future, we can look at a benchmark targeting that usecase.

Reply:
> That says more about the flakiness of these nanobenchmarks than about the added utility of the constant.

Top of the manul of BenchmarkTools.jl:

> The reasoning behind our definition of "sample" may not be obvious to all readers. If the time to execute a benchmark is smaller than the resolution of your timing method, then a single evaluation of the benchmark will generally not produce a valid sample. In that case, one must approximate a valid sample by recording the total time t it takes to record n evaluations, and estimating the sample's time per evaluation as t/n. For example, if a sample takes 1 second for 1 million evaluations, the approximate time per evaluation for that sample is 1 microsecond. It's not obvious what the right number of evaluations per sample should be for any given benchmark, so BenchmarkTools provides a mechanism (the tune! method) to automatically figure it out for you.

your benchmark is less than the resolution of the timing method, and so producing a junk result.

Reply:
Yes, I'm aware of that. I'm still missing a motivating actual usecase that's not a synthetic microbenchmark just calling `hash` in a loop.

Reply:
I'm not "purposefully writing a slower version":

<details>
<summary>Code output</summary>

```

julia> g(x::Some, h) = hash(x.value, hash(_hashSome, h))
g (generic function with 2 methods)

julia> f(x::Some, h) = hash(x.value, hash(Some, h))
f (generic function with 2 methods)

julia> code_llvm(g, Tuple{Some{Int}, UInt}; debuginfo=:none)
; Function Signature: g(Base.Some{Int64}, UInt64)
define i64 @julia_g_833([1 x i64]* nocapture noundef nonnull readonly align 8 dereferenceable(8) %"x::Some", i64 zeroext %"h::UInt64") #0 {
top:
  %"x::Some.value_ptr" = getelementptr inbounds [1 x i64], [1 x i64]* %"x::Some", i64 0, i64 0
  %bitcast = load i64, i64* %"x::Some.value_ptr", align 8
  %0 = xor i64 %bitcast, -1
  %1 = shl i64 %bitcast, 21
  %2 = add i64 %1, %0
  %3 = lshr i64 %2, 24
  %4 = xor i64 %3, %2
  %5 = mul i64 %4, 265
  %6 = lshr i64 %5, 14
  %7 = xor i64 %6, %5
  %8 = mul i64 %7, 21
  %9 = lshr i64 %8, 28
  %10 = xor i64 %9, %8
  %11 = mul i64 %10, 2147483649
  %.neg = mul i64 %"h::UInt64", 9
  %.neg17 = add i64 %.neg, -1323963221037668862
  %12 = add i64 %.neg17, %11
  ret i64 %12
}

julia> code_llvm(f, Tuple{Some{Int}, UInt}; debuginfo=:none)
; Function Signature: f(Base.Some{Int64}, UInt64)
define i64 @julia_f_908([1 x i64]* nocapture noundef nonnull readonly align 8 dereferenceable(8) %"x::Some", i64 zeroext %"h::UInt64") #0 {
top:
  %"x::Some.value_ptr" = getelementptr inbounds [1 x i64], [1 x i64]* %"x::Some", i64 0, i64 0
  %0 = mul i64 %"h::UInt64", 3
  %1 = call i64 bitcast (void ()* @jlplt_jl_type_hash_915_got.jit to i64 ({}*)*)({}* @"jl_global#913.jit")
  %2 = sub i64 %0, %1
  %bitcast = load i64, i64* %"x::Some.value_ptr", align 8
  %3 = insertelement <2 x i64> poison, i64 %2, i64 0
  %4 = insertelement <2 x i64> %3, i64 %bitcast, i64 1
  %5 = xor <2 x i64> %4, <i64 -1, i64 -1>
  %6 = shl <2 x i64> %4, <i64 21, i64 21>
  %7 = add <2 x i64> %6, %5
  %8 = lshr <2 x i64> %7, <i64 24, i64 24>
  %9 = xor <2 x i64> %8, %7
  %10 = mul <2 x i64> %9, <i64 265, i64 265>
  %11 = lshr <2 x i64> %10, <i64 14, i64 14>
  %12 = xor <2 x i64> %11, %10
  %13 = mul <2 x i64> %12, <i64 21, i64 21>
  %14 = lshr <2 x i64> %13, <i64 28, i64 28>
  %15 = xor <2 x i64> %14, %13
  %16 = mul <2 x i64> %15, <i64 -6442450947, i64 2147483649>
  %shift = shufflevector <2 x i64> %16, <2 x i64> poison, <2 x i32> <i32 1, i32 undef>
  %17 = add <2 x i64> %16, %shift
  %18 = extractelement <2 x i64> %17, i64 0
  ret i64 %18
}

julia> code_native(g, Tuple{Some{Int}, UInt}; debuginfo=:none)
	.text
	.file	"g"
	.globl	julia_g_1004                    # -- Begin function julia_g_1004
	.p2align	4, 0x90
	.type	julia_g_1004,@function
julia_g_1004:                           # @julia_g_1004
; Function Signature: g(Base.Some{Int64}, UInt64)
# %bb.0:                                # %top
	#DEBUG_VALUE: g:x <- [DW_OP_deref] [$rdi+0]
	#DEBUG_VALUE: g:h <- $rsi
	push	rbp
	mov	rax, qword ptr [rdi]
	lea	rdx, [rsi + 8*rsi]
	mov	rbp, rsp
	mov	rcx, rax
	not	rcx
	shl	rax, 21
	add	rax, rcx
	mov	rcx, rax
	shr	rcx, 24
	xor	rcx, rax
	imul	rax, rcx, 265
	mov	rcx, rax
	shr	rcx, 14
	xor	rcx, rax
	lea	rax, [rcx + 4*rcx]
	lea	rax, [rcx + 4*rax]
	mov	rcx, rax
	shr	rcx, 28
	xor	rcx, rax
	mov	rax, rcx
	shl	rax, 31
	add	rax, rcx
	add	rdx, rax
	movabs	rax, -1323963221037668862
	add	rax, rdx
	pop	rbp
	ret
.Lfunc_end0:
	.size	julia_g_1004, .Lfunc_end0-julia_g_1004
                                        # -- End function
	.section	".note.GNU-stack","",@progbits

julia> code_native(f, Tuple{Some{Int}, UInt}; debuginfo=:none)
	.text
	.file	"f"
	.section	.rodata.cst8,"aM",@progbits,8
	.p2align	3                               # -- Begin function julia_f_1081
.LCPI0_0:
	.quad	265                             # 0x109
.LCPI0_1:
	.quad	21                              # 0x15
	.section	.rodata.cst16,"aM",@progbits,16
	.p2align	4
.LCPI0_2:
	.quad	-6442450947                     # 0xfffffffe7ffffffd
	.quad	2147483649                      # 0x80000001
	.text
	.globl	julia_f_1081
	.p2align	4, 0x90
	.type	julia_f_1081,@function
julia_f_1081:                           # @julia_f_1081
; Function Signature: f(Base.Some{Int64}, UInt64)
# %bb.0:                                # %top
	#DEBUG_VALUE: f:x <- [DW_OP_deref] [$rdi+0]
	#DEBUG_VALUE: f:h <- $rsi
	push	rbp
	mov	rbp, rsp
	push	r14
	push	rbx
	mov	r14, rdi
	movabs	rdi, offset ".Ljl_global#1086.jit"
	movabs	rax, offset .Ljlplt_jl_type_hash_1088_got.jit
	lea	rbx, [rsi + 2*rsi]
	call	rax
	vmovq	xmm1, qword ptr [r14]           # xmm1 = mem[0],zero
	sub	rbx, rax
	movabs	rax, offset .LCPI0_0
	vmovq	xmm0, rbx
	vpunpcklqdq	xmm0, xmm0, xmm1        # xmm0 = xmm0[0],xmm1[0]
	vpsllq	xmm1, xmm0, 21
	vpternlogq	xmm0, xmm0, xmm0, 15
	vpaddq	xmm0, xmm1, xmm0
	vpsrlq	xmm1, xmm0, 24
	vpxor	xmm0, xmm1, xmm0
	vpmullq	xmm0, xmm0, qword ptr [rax]{1to2}
	movabs	rax, offset .LCPI0_1
	vpsrlq	xmm1, xmm0, 14
	vpxor	xmm0, xmm1, xmm0
	vpmullq	xmm0, xmm0, qword ptr [rax]{1to2}
	movabs	rax, offset .LCPI0_2
	vpsrlq	xmm1, xmm0, 28
	vpxor	xmm0, xmm1, xmm0
	vpmullq	xmm0, xmm0, xmmword ptr [rax]
	vpshufd	xmm1, xmm0, 238                 # xmm1 = xmm0[2,3,2,3]
	vpaddq	xmm0, xmm0, xmm1
	vmovq	rax, xmm0
	pop	rbx
	pop	r14
	pop	rbp
	ret
.Lfunc_end0:
	.size	julia_f_1081, .Lfunc_end0-julia_f_1081
                                        # -- End function
.set ".Ljl_global#1086.jit", 140529893066128
	.size	".Ljl_global#1086.jit", 0
	.type	.Ljlplt_jl_type_hash_1088_got.jit,@function
.set .Ljlplt_jl_type_hash_1088_got.jit, 140530174606400
	.section	".note.GNU-stack","",@progbits
```

</details>

These need to do the same amount of work in the end; the hashing of the `Some` type ends up vectorized with the hashing of the e.g. input integer in this best case scenario. For bigger objects in that `Some`, the hash of the wrapped object will be the bottleneck. I just don't think it's worth the maintenance overhead to add that specialization & document why it's there for a measly 4ns that, so far, only show up in a synthetic benchmark.

If anything, this seems to indicate a failure of the compiler to inline that constant, as a type can't be deleted. This is because hashing a type does a call to `jl_type_hash`; probably just something that got missed. The hash of a type is already cached in the type object itself after all:

```
julia> ccall(:jl_type_hash, UInt, (Any,), Some{Int})
0x00000000d45c7293

julia> Some{Int}.<tab>
flags       hash
instance    layout
name        parameters
super       types

julia> Some{Int}.hash % UInt
0xffffffffd45c7293
```

Reply:
Doesn't vectorize on my machine:
```julia
julia> code_llvm(g, Tuple{Some{Int}, UInt}; debuginfo=:none)
; Function Signature: g(Base.Some{Int64}, UInt64)
define i64 @julia_g_2497([1 x i64]* nocapture noundef nonnull readonly align 8 dereferenceable(8) %"x::Some", i64 zeroext %"h::UInt64") #0 {
top:
  %"x::Some.value_ptr" = getelementptr inbounds [1 x i64], [1 x i64]* %"x::Some", i64 0, i64 0
  %bitcast = load i64, i64* %"x::Some.value_ptr", align 8
  %0 = xor i64 %bitcast, -1
  %1 = shl i64 %bitcast, 21
  %2 = add i64 %1, %0
  %3 = lshr i64 %2, 24
  %4 = xor i64 %3, %2
  %5 = mul i64 %4, 265
  %6 = lshr i64 %5, 14
  %7 = xor i64 %6, %5
  %8 = mul i64 %7, 21
  %9 = lshr i64 %8, 28
  %10 = xor i64 %9, %8
  %11 = mul i64 %10, 2147483649
  %.neg = mul i64 %"h::UInt64", 9
  %.neg17 = add i64 %.neg, -1931804474194194988
  %12 = add i64 %.neg17, %11
  ret i64 %12
}

julia> code_llvm(f, Tuple{Some{Int}, UInt}; debuginfo=:none)
; Function Signature: f(Base.Some{Int64}, UInt64)
define i64 @julia_f_2509([1 x i64]* nocapture noundef nonnull readonly align 8 dereferenceable(8) %"x::Some", i64 zeroext %"h::UInt64") #0 {
top:
  %"x::Some.value_ptr" = getelementptr inbounds [1 x i64], [1 x i64]* %"x::Some", i64 0, i64 0
  %0 = mul i64 %"h::UInt64", 3
  %1 = call i64 bitcast (void ()* @jlplt_jl_type_hash_2516_got.jit to i64 ({}*)*)({}* @"jl_global#2514.jit")
  %2 = sub i64 %0, %1
  %3 = xor i64 %2, -1
  %4 = shl i64 %2, 21
  %5 = add i64 %4, %3
  %6 = lshr i64 %5, 24
  %7 = xor i64 %6, %5
  %8 = mul i64 %7, 265
  %9 = lshr i64 %8, 14
  %10 = xor i64 %9, %8
  %11 = mul i64 %10, 21
  %12 = lshr i64 %11, 28
  %13 = xor i64 %12, %11
  %bitcast = load i64, i64* %"x::Some.value_ptr", align 8
  %14 = xor i64 %bitcast, -1
  %15 = shl i64 %bitcast, 21
  %16 = add i64 %15, %14
  %17 = lshr i64 %16, 24
  %18 = xor i64 %17, %16
  %19 = mul i64 %18, 265
  %20 = lshr i64 %19, 14
  %21 = xor i64 %20, %19
  %22 = mul i64 %21, 21
  %23 = lshr i64 %22, 28
  %24 = xor i64 %23, %22
  %25 = mul i64 %24, 2147483649
  %.neg = mul i64 %13, -6442450947
  %26 = add i64 %.neg, %25
  ret i64 %26
}
```

Reply:
I do have AVX512, which may be required. Either way, the compiler should be able to constant fold the hash of the type into the function, so the point is IMO moot.

Reply:
> hash needs to obey x == y => hash(x) == hash(y), so I believe this change also needs to modify hash.

`hash` only needs to obey `isequal`, as it turns out; since `isequal` defaults to `==`, having the custom hashing is better either way.